server/src/main/kotlin/org/jetbrains/bsp/bazel/server/sync/languages/rust/RustLanguagePlugin.kt

```diff
@@ -0,0 +1,122 @@
+        targetInfo.getRustCrateInfoOrNull()?.let { rustCrateInfo ->
+            val location = resolveTargetLocation(rustCrateInfo)
+            val crateRoot = resolveTargetCrateRoot(rustCrateInfo, location)
+
+            RustModule(
+                crateId = rustCrateInfo.crateId,
+                location = location,
+                fromWorkspace = rustCrateInfo.fromWorkspace,
+                name = rustCrateInfo.name,
+                kind = rustCrateInfo.kind,
+                edition = rustCrateInfo.edition,
+                crateFeatures = rustCrateInfo.crateFeaturesList,
+                dependenciesCrateIds = rustCrateInfo.dependenciesCrateIdsList,
+                crateRoot = crateRoot,
+                version = rustCrateInfo.version,
+                procMacroArtifacts = rustCrateInfo.procMacroArtifactsList,
+            )
+        }
+
+    private fun resolveTargetLocation(rustCrateInfo: RustCrateInfo): RustCrateLocation =
+        when (rustCrateInfo.location) {
+            BspTargetInfo.RustCrateLocation.WORKSPACE_DIR -> RustCrateLocation.WORKSPACE_DIR
+            else -> RustCrateLocation.EXEC_ROOT
+        }
+
+    private fun resolveTargetCrateRoot(rustCrateInfo: RustCrateInfo, location: RustCrateLocation): String {
+        val path = if (location == RustCrateLocation.WORKSPACE_DIR) {
+            bazelPathsResolver.relativePathToWorkspaceAbsolute(rustCrateInfo.crateRoot)
+        } else {
+            bazelPathsResolver.relativePathToExecRootAbsolute(rustCrateInfo.crateRoot)
+        }
+
+        return bazelPathsResolver.resolveUri(path).toString()
+    }
+
+    fun toRustWorkspaceResult(requestTargets: List<Module>, allTargets: List<Module>): RustWorkspaceResult {
+        val modules = findAllRelatedRustTargets(requestTargets, allTargets.associateBy { it.label })
+        val packages = rustPackageResolver.rustPackages(modules)
+        val (dependencies, rawDependencies) = rustDependencyResolver.rustDependencies(packages, modules)
+        val resolvedTargets = packages.filter { it.origin == "WORKSPACE" }
+                .flatMap { it.resolvedTargets.map { it2 -> it.id + ':' + it2.name }}
+                .map { BuildTargetIdentifier(it) }
+
+        return RustWorkspaceResult(
+            packages,
+            rawDependencies,
+            dependencies,
+            resolvedTargets
+        )
+    }
+
+    /**
+     * Find all targets that are needed to be resolved.
+     * It includes all targets from `targets` and all their dependencies.
+     *
+     * Let's assume we have a following project:
+     *           B    A
+     *           |   / \
+     *           C  D   E
+     *           \ /    |
+     *            F     G
+     * We want to resolve all targets related to B. We only need B, C and F.
+     *
+     * There is no need to revisit already visited targets - we can use `visited` set for that.
+     * */
+    private fun findAllRelatedRustTargets(
```

Comment:
cant we reuse DependencyTree here?

Reply:
bump, if not it should be a tailrec

Reply:
It would be possible, but I think that it would create too much unnecessary conversion comparing to what we have now. DependencyTree operates on info from TargetInfo and here we operate on Modules and Labels and to fit the needs of DependencyTree we would have to convert everything and then reconvert it again to Modules. So I didn't use DependencyTree but I changed the function to be tailrec